packages/firestore/src/remote/watch_change.ts

```diff
@@ -350,35 +222,318 @@ export class WatchChangeAggregator {
   }
 
+  /** Resets a target after an existence filter mismatch. */
+  handleExistenceFilterMismatch(targetId: TargetId): void {
+    this.resetTarget(targetId);
+  }
+
+  /**
+   * Converts the currently accumulated state into a remote event at the
+   * provided snapshot version. Resets the accumulated changes before returning.
+   */
+  createRemoteEvent(snapshotVersion: SnapshotVersion): RemoteEvent {
+    const targetChanges: { [targetId: number]: TargetChange } = {};
+
+    objUtils.forEachNumber(this.targetStates, (targetId, targetState) => {
+      if (this.isActiveTarget(targetId)) {
+        const queryData = this.queryDataCallback(targetId);
+
+        let addedDocuments = documentKeySet();
+        let modifiedDocuments = documentKeySet();
+        let removedDocuments = documentKeySet();
+
+        if (targetState.snapshotChanges.size !== 0) {
+          targetState.snapshotChanges.forEach((key, changeType) => {
+            switch (changeType) {
+              case ChangeType.Added:
+                addedDocuments = addedDocuments.add(key);
+                break;
+              case ChangeType.Modified:
+                modifiedDocuments = modifiedDocuments.add(key);
+                break;
+              case ChangeType.Removed:
+                removedDocuments = removedDocuments.add(key);
+                break;
+              default:
+                fail('Encountered invalid change type: ' + changeType);
+            }
+          });
+        } else if (targetState.current && queryData.query.isDocumentQuery()) {
+          // Document queries for document that don't exist can produce an empty
+          // result set. To update our local cache, we synthesize a document
+          // delete if we have not previously received the document. This
+          // resolves the limbo state of the document, removing it from
+          // limboDocumentRefs.
+          //
+          // TODO(dimond): Ideally we would have an explicit lookup query
+          // instead resulting in an explicit delete message and we could
+          // remove this special logic.
+          const key = new DocumentKey(queryData.query.path);
+          if (!this.hasSyncedDocument(targetId, key)) {
+            this.documentUpdates = this.documentUpdates.insert(
+              key,
+              new NoDocument(key, snapshotVersion)
+            );
+
+            // While we don't add the document to a target, we potentially
+            // need to mark it as a resolved limbo key. This requires us
+            // to add the document to the list of document target mappings.
+            this.ensureDocumentTargetMapping(key);
+          }
+        }
+
+        targetChanges[targetId] = {
+          current: targetState.current,
+          resumeToken: targetState.resumeToken,
+          snapshotVersion,
+          addedDocuments,
+          modifiedDocuments,
+          removedDocuments
+        };
+
+        targetState.snapshotChanges = snapshotChangesMap();
```

Comment:
Two other options for how to avoid `CurrentStatusUpdate`:

  * Keep a `lastCurrent` in `TargetState` and promote it here
  * Make an `existingCurrent` callback similar to the `existingKeys` mechanism you added for plumbing in existing prior key state.

Reply:
`shouldRaise` is essentially option one, but it also takes into account resume tokens and document updates.